tools/CheckH5.py

```diff
@@ -0,0 +1,207 @@
+import os
+import h5py
+
+
+# Describes a particular comparison between H5 datasets or groups
+class H5Check:
+    # `unit_test` must be the unit test object -- this class is to be
+    # constructed within a test case, and must have access to the
+    # calling unit test object so that it can access the assert and
+    # subTest member functions.
+    def __init__(self, unit_test, test_h5_label, test_h5_entity,
+                 expected_h5_entity, absolute_tolerance, relative_tolerance):
+        self.unit_test = unit_test
+        self.test_h5_label = test_h5_label
+        self.test_h5_entity = test_h5_entity
+        self.expected_h5_entity = expected_h5_entity
+        self.absolute_tolerance = float(absolute_tolerance)
+        self.relative_tolerance = (None if relative_tolerance is None else
+                                   float(relative_tolerance))
+
+    # True if `test data` is approximately equal to `expected_data`, otherwise
+    # False. It is expected that both inputs to the function are types for
+    # which `abs` returns a float
+    def approx(self, test_data, expected_data):
+        return abs(test_data - expected_data) < self.absolute_tolerance and (
+            max(abs(test_data), abs(expected_data)) == 0.0 or
+            self.relative_tolerance is None or abs(test_data - expected_data) /
+            max(abs(test_data), abs(expected_data)) < self.relative_tolerance)
+
+    # Perform the unit test comparisons between the dataset or group
+    # represented by the string `test_entity` and `expected_entity`.
+    # If `expected_entity` is `None`, then `test_entity` (or, if it is a group,
+    # the datasets within the group) is compared to 0.0.
+    def check_h5_file(self, h5_file, test_entity, expected_entity):
+        if isinstance(h5_file[test_entity], h5py.Dataset):
+            # It is recommended to include subTest commands once python3 can be
+            # required:
+            # with self.unit_test.subTest(test_entity=test_entity,
+            #                             expected_entity=expected_entity):
+            test_data = h5_file[test_entity][()]
+            if self.expected_h5_entity is not None:
+                expected_data = h5_file[expected_entity][()]
+                self.unit_test.assertEqual(
+                    test_data.shape, expected_data.shape,
+                    "test and expected h5 datasets must have identical"
+                    " sizes.")
+                self.unit_test.assertEqual(
+                    test_data.dtype, expected_data.dtype,
+                    "test and expected h5 datasets must have identical"
+                    " types.")
+                if (test_data.dtype == np.dtype(np.float)
+                        or test_data.dtype == np.dtype(np.complexfloating)):
+                    for index in np.ndindex(test_data.shape):
+                        # It is recommended to include subTest commands once
+                        # python3 can be required:
```

Comment:
what we would do with subtests?  Could you elaborate a bit more on what exactly this will do?

Reply:
The subtests would ensure that:
- any failing test would be annotated with the dataset and index that failed
- the tests would continue after a failing item in the dataset was found

Reply:
:+1: I like that :)